### ISSUE TITLE
Exception on the video player after the app opens from the background 

### ISSUE THREAD

**Issue by ziperjohn** [flutter/flutter#154602]
### What package does this bug report belong to?

video_player

### What target platforms are you seeing this bug on?

Android

### Have you already upgraded your packages?

Yes

### Dependency versions

<details><summary>pubspec.lock</summary>

```lock
[Paste file content here]
```

</details>

### Steps to reproduce

1. Create new flutter app
2. Paste code sample
3. Run app on Android
4. Video is playing now,
5. Put the app in the background 
6. Open app from the background
7. An exception occurs

### Expected results

The video will start playing again without exception.

I tested this situation on iOS and work fine.

### Actual results

After open app from the background an exception occurs. 

[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Bad state: Future already completed
E/flutter (17608): #0      _AsyncCompleter.complete (dart:async/future_impl.dart:43:31)
E/flutter (17608): #1      VideoPlayerController.initialize.eventListener (package:video_player/video_player.dart:468:33)
E/flutter (17608): #2      _RootZone.runUnaryGuarded (dart:async/zone.dart:1594:10)
E/flutter (17608): #3      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:365:11)
E/flutter (17608): #4      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:297:7)
E/flutter (17608): #5      _ForwardingStreamSubscription._add (dart:async/stream_pipe.dart:123:11)
E/flutter (17608): #6      _MapStream._handleData (dart:async/stream_pipe.dart:218:10)
E/flutter (17608): #7      _ForwardingStreamSubscription._handleData (dart:async/stream_pipe.dart:153:13)
E/flutter (17608): #8      _RootZone.runUnaryGuarded (dart:async/zone.dart:1594:10)
E/flutter (17608): #9      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:365:11)
E/flutter (17608): #10     _DelayedData.perform (dart:async/stream_impl.dart:541:14)
E/flutter (17608): #11     _PendingEvents.handleNext (dart:async/stream_impl.dart:646:11)
E/flutter (17608): #12     _PendingEvents.schedule.<anonymous closure> (dart:async/stream_impl.dart:617:7)
E/flutter (17608): #13     _microtaskLoop (dart:async/schedule_microtask.dart:40:21)
E/flutter (17608): #14     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49:5)

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:video_player/video_player.dart';

void main() {
  runApp(const App());
}

class App extends StatelessWidget {
  const App({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: VideoPlayerScreen(),
    );
  }
}

class VideoPlayerScreen extends StatefulWidget {
  const VideoPlayerScreen({super.key});

  @override
  State<VideoPlayerScreen> createState() => _VideoPlayerScreenState();
}

class _VideoPlayerScreenState extends State<VideoPlayerScreen> {
  late VideoPlayerController _controller;

  @override
  void initState() {
    super.initState();

    _controller =
        VideoPlayerController.networkUrl(Uri.parse('https://flutter.github.io/assets-for-api-docs/assets/videos/bee.mp4'))
          ..initialize().then((_) {
            // Ensure the first frame is shown after the video is initialized, even before the play button has been pressed.
            setState(() {});
          });

    _controller.setLooping(true);
    _controller.play();
  }

  @override
  void dispose() {
    super.dispose();

    _controller.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("video player"),
      ),
      body: Center(
        child: _controller.value.isInitialized
            ? AspectRatio(
                aspectRatio: _controller.value.aspectRatio,
                child: VideoPlayer(_controller),
              )
            : Container(),
      ),
    );
  }
}

```

</details>

### Screenshots or Videos

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.24.0, on macOS 14.6.1 23G93 darwin-arm64, locale en-CZ)
    • Flutter version 3.24.0 on channel stable at /Users/janvaverka/Development Tools/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 80c2e84975 (5 weeks ago), 2024-07-30 23:06:49 +0700
    • Engine revision b8800d88be
    • Dart version 3.5.0
    • DevTools version 2.37.2

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at /Users/janvaverka/Library/Android/sdk
    • Platform android-35, build-tools 33.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.92.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.96.0

[✓] Connected device (4 available)
    • SM A528B (mobile)               • RZCT101GZRW           • android-arm64  • Android 14 (API 34)
    • macOS (desktop)                 • macos                 • darwin-arm64   • macOS 14.6.1 23G93 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         • macOS 14.6.1 23G93 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript • Google Chrome 128.0.6613.114
    ! Error: Browsing on the local area network for Ziper. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
    ! Error: Browsing on the local area network for Ziper. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by qinjing100** [comment]
I also encounter the problem in debug mode in Android 11, Android 14.
1. I play the video from local file
2. bring app to background
3. bring app foreground again
4. Bad state: Future already completed

```
I/flutter (10951): ----------------FIREBASE CRASHLYTICS----------------
I/flutter (10951): Bad state: Future already completed
I/flutter (10951): #0      _AsyncCompleter.complete (dart:async/future_impl.dart:43:31)
I/flutter (10951): #1      VideoPlayerController.initialize.eventListener (package:video_player/video_player.dart:468:33)
I/flutter (10951): #2      _RootZone.runUnaryGuarded (dart:async/zone.dart:1594:10)
I/flutter (10951): #3      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:365:11)
I/flutter (10951): #4      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:297:7)
I/flutter (10951): #5      _ForwardingStreamSubscription._add (dart:async/stream_pipe.dart:123:11)
I/flutter (10951): #6      _MapStream._handleData (dart:async/stream_pipe.dart:218:10)
I/flutter (10951): #7      _ForwardingStreamSubscription._handleData (dart:async/stream_pipe.dart:153:13)
I/flutter (10951): #8      _RootZone.runUnaryGuarded (dart:async/zone.dart:1594:10)
I/flutter (10951): #9      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:365:11)
I/flutter (10951): #10     _DelayedData.perform (dart:async/stream_impl.dart:541:14)
I/flutter (10951): #11     _PendingEvents.handleNext (dart:async/stream_impl.dart:646:11)
I/flutter (10951): #12     _PendingEvents.schedule.<anonymous closure> (dart:async/stream_impl.dart:617:7)
I/flutter (10951): #13     _micro
I/flutter (10951): ----------------------------------------------------
W/FirebaseCrashlytics(10951): Unable to read App Quality Sessions session id.
Restarted application in 3,772ms.

```

Reproduced on video_player: ^2.9.1 and Flutter stable 3.24.0

```
[✓] Flutter (Channel stable, 3.24.0, on macOS 14.6.1 23G93 darwin-x64, locale zh-Hans-HK)
    • Flutter version 3.24.0 on channel stable at /Users/bruce/fvm/versions/3.24.0
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 80c2e84975 (5 weeks ago), 2024-07-30 23:06:49 +0700
    • Engine revision b8800d88be
    • Dart version 3.5.0
    • DevTools version 2.37.2
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/bruce/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15E204a
    ✗ Unable to get list of installed Simulator runtimes.
    ! CocoaPods 1.11.3 out of date (1.13.0 is recommended).
        CocoaPods is a package manager for iOS or macOS platform code.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/to/platform-plugins
      To update CocoaPods, see https://guides.cocoapods.org/using/getting-started.html#updating-cocoapods

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] VS Code (version 1.92.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.94.0

[!] Proxy Configuration
    • HTTP_PROXY is set
    ! NO_PROXY is not set

[✓] Connected device (3 available)
    • V2164A (mobile) • 344393009600178 • android-arm64  • Android 11 (API 30)
    • macOS (desktop) • macos           • darwin-x64     • macOS 14.6.1 23G93 darwin-x64
    • Chrome (web)    • chrome          • web-javascript • Google Chrome 128.0.6613.114

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.

```


this problem not occur in Android 9.

Please help fix this issue, thanks a lot.

**Comment by msayed-net** [comment]
Same issue here

**Comment by reidbaker** [comment]
@ziperjohn @msayed-net is this bug a regression? if so what version of video player works correctly (so we can bisect)

**Comment by michael-joseph-payne** [comment]
I see this as well, but with a streaming (network) video.

**Comment by qinjing100** [comment]
I found with flutter version 3.22.3 it works well, the problem not occurs.
I think of there is something bug of  _VideoAppLifeCycle (video_player 2.9.1) with flutter sdk (3.24.0 +).
something like issue #154559 .

**Comment by dfdgsdfg** [comment]
Workaround.

tldr;
```
# pubspec.yaml

dependencies:
  video_player: ^2.9.1

dependency_overrides:
  video_player_android: 2.7.1
```

https://github.com/flutter/flutter/issues/154559#issuecomment-2342626756

**Comment by michael-joseph-payne** [comment]
How is this closed or completed?

**Comment by huycozy** [comment]
Re-open this issue since the working version (`video_player_android: 2.7.1`) seems to have been provided at https://github.com/flutter/flutter/issues/154602#issuecomment-2342632947. 

Based on video_player_android changelog, this may be related to Impeller support: https://pub.dev/packages/video_player_android/changelog#271

**Comment by matanlurey** [comment]
I can take a look, though I'm suspecting that it's fixed in [2.7.4](https://pub.dev/packages/video_player_android/changelog#274) already.

**Comment by matanlurey** [comment]
I can reproduce the error. I don't think it's related to Impeller but it could be related to the initialization logic we changed.

**Comment by matanlurey** [comment]
Ok, here is a summary of the issue:

```dart
// video_player.dart
void eventListener(VideoEvent event) {
  if (_isDisposed) {
    return;
  }

  switch (event.eventType) {
    case VideoEventType.initialized:
      value = value.copyWith(
        duration: event.duration,
        size: event.size,
        rotationCorrection: event.rotationCorrection,
        isInitialized: event.duration != null,
        errorDescription: null,
        isCompleted: false,
      );
      initializingCompleter.complete(null);
```

There is an implicit contract that `VideoEventType.initialized` is invoked exactly once; however that is no longer true when a video is released (background resource) and recreated. It's not clear (to me) what the right fix is here - personally I think the logic is faulty assuming that a video is only initialized once, but that's a larger breaking change.

Instead, I'll adjust `video_player_android` not to send the event twice for the same player.

**Comment by lhlyu** [comment]
Tried it, no issues with `2.7.1`, but it doesn’t work with `2.7.5`.

**Comment by michael-joseph-payne** [comment]
Still seen in:

video_player: ^2.9.2

So far, just a Pixel 8 Pro.

```
Exception - This event
error
03:39:27.000
StateError: Bad state: VideoPlayerController already initialized
Ui Lifecycle
info
03:39:24.913

{
screen: MainActivity,
state: resumed
}
App Lifecycle
info
03:39:24.912

{
state: foreground
}
Ui Lifecycle
info
03:39:24.910

{
screen: MainActivity,
state: started
}
Device Event
info
03:39:24.830

{
action: DREAMING_STOPPED
}
Device Event
info
03:39:24.220

{
action: SCREEN_ON
}
App Lifecycle
info
03:39:15.902

{
state: background
}
Device Event
info
03:39:15.496

{
action: SCREEN_OFF
}
Device Event
info
03:39:15.420

{
action: DREAMING_STARTED
}
Ui Lifecycle
info
03:39:15.348

{
screen: MainActivity,
state: saveInstanceState
}
Ui Lifecycle
info
03:39:15.295

{
screen: MainActivity,
state: stopped
}
Ui Lifecycle
info
03:39:15.198

{
screen: MainActivity,
state: paused
}
Network Event
info
03:38:23.381

{
action: NETWORK_CAPABILITIES_CHANGED,
download_bandwidth: 53520,
network_type: wifi,
signal_strength: -40,
upload_bandwidth: 38976,
vpn_active: false
}
Network Event
info
03:38:17.047

{
action: NETWORK_CAPABILITIES_CHANGED,
download_bandwidth: 69288,
network_type: wifi,
signal_strength: -40,
upload_bandwidth: 38976,
vpn_active: false
}
Navigation
info
03:37:02.614

{
state: didPop,
to: /
}
Navigation
info
03:37:01.277

{
from: /,
state: didPush
}
Navigation
info
03:36:58.308

{
from: /sign_in,
state: didPop,
to: /
}

**Comment by michael-joseph-payne** [comment]
It looks like this was fixed in video_player_android 2.7.6

I imagine anyone with this issue would want to use the following:

dependency_overrides:
  video_player_android: 2.7.6

I'm currently testing and will feedback if I still see this.

**Comment by gnprice** [comment]
Knowing which release an issue was fixed in is useful information for anyone who has the issue and finds the thread. (Especially for issues fixed in packages, rather than in Flutter itself.) That information wasn't on this thread before, so thanks @michael-joseph-payne for adding it here.
